packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -120,22 +174,29 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     internal
     override
     returns (bytes memory context, bool revertOnRecipientRevert) {
-        (signature);
-        require(approvalData.length == 0, "approvalData: invalid length");
-        if (relayRequest.relayData.paymasterData.length != 0) {
-            require(relayRequest.relayData.paymasterData.length >= 4, "paymastaData: missing method sig");
+        (signature, approvalData);
+
+        // paymasterData must contain the token address, and optionally a a valid "permit" call on the token.
+        require(relayRequest.relayData.paymasterData.length >= 20, "must contain token address");
+        IERC20Metadata token = _getTokenFromPaymasterData(relayRequest.relayData.paymasterData);
+        require(isTokenSupported(token),"unsupported token");
+        if (relayRequest.relayData.paymasterData.length != 20) {
+            require(relayRequest.relayData.paymasterData.length >= 24, "must contain \"permit\" and token");
             require(
-                permitMethodSignature == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
-                "paymasterData: wrong method sig");
+                permitMethodSignatures[token] == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
+                "wrong \"permit\" method sig");
             // execute permit method for this token
             // solhint-disable-next-line avoid-low-level-calls
-            (bool success, bytes memory ret) = address(token).call(relayRequest.relayData.paymasterData);
+            (bool success, bytes memory ret) = address(token).call(relayRequest.relayData.paymasterData[:relayRequest.relayData.paymasterData.length - 20]);
```

Comment:
your paymasterData format is very strange:
either "permit" followed by token, 
or only "token".
better have fixed offsets (e.g. token at offset 0, permit starts at offset 20)

Reply:
You baiscally tell me to swap their position. I can do that, but what do we gain from it?

Reply:
I tend to agree with Dror and TypeScript here - optional fields should be in the end. I cannot explain why but this feels like this can make sense.

Reply:
Fixed